### PR TITLE
PR: Add support for LSP workspace calls

### LINKED ISSUE #9007: Connect our LSP client to the workspace functionality provided by the PyLS.

### PR DESCRIPTION
**Pull request by andfoy**
## Description of Changes
We've enabled support for LSP workspace capabilities, including changing workspace folders on projects and single files, as well as adding support for simultaneous workspaces. This PR also enables symbol searching.

* [x] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #9007

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
@andfoy 

---
**TODO tasks**
- [x] Handle ``workspace/workspaceFolders`` call
- [x] Send ``workspace/didChangeWorkspaceFolders``
- [x] Handle ``workspace/configuration`` request
- [x] Send ``workspace/didChangeWatchedFiles`` request
- [x] Send ``workspace/symbol`` request
- [x] Connect ``workspace/didChangeWorkspaceFolders`` with project file watcher
~- [ ] Define a workspace client per registered project.~
~- [ ] Connect symbol search in project with find in files widget~

### CODE REVIEW

File: spyder/plugins/editor/lsp/providers/workspace.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# -*- coding: utf-8 -*-
+
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""Spyder Language Server Protocol Client window handler routines."""
```
Comment (ccordoba12) [inline spyder/plugins/editor/lsp/providers/workspace.py]: Please update this docstring.

File: spyder/plugins/editor/lsp/providers/workspace.py [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+    def send_execute_command(self, params):
+        # It is not clear how this call is invoked
+        params = {
+            'command': params['command'],
+            'arguments': params['args']
+        }
+        return params
+
+    @handles(LSPRequestTypes.WORKSPACE_APPLY_EDIT)
+    def apply_edit(self, response):
```
Comment (andfoy) [inline spyder/plugins/editor/lsp/providers/workspace.py]: @ccordoba12 Shall we implement this call in another PR?

**Comment by andfoy** [comment]
@ccordoba12 I think all the client calls are implemented, some Spyder implementations are missing, such as ``workspace/symbol`` and ``workspace/applyEdit``, which should be implemented in follow-up PRs. Other functions such as ``workspace/executeCommand`` depend on other features, such as Code Lens and Code Actions, therefore this call depends on implementing the aforementioned features.

File: spyder/plugins/editor/lsp/manager.py [hunk 3]
```diff
@@ -240,6 +244,14 @@ def send_request(self, language, request, params):
                 client = self.clients[language]['instance']
                 client.perform_request(request, params)
 
+    def broadcast_request(self, request, params):
```
Comment (ccordoba12) [inline spyder/plugins/editor/lsp/manager.py]: Please add a docstring to this method.

File: spyder/plugins/editor/lsp/providers/workspace.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    @send_notification(method=LSPRequestTypes.WORKSPACE_CONFIGURATION_CHANGE)
+    def send_plugin_configurations(self, configurations, *args):
+        self.plugin_configurations = configurations
+        params = {
+            'settings': configurations
+        }
+        return params
+
+    @handles(LSPRequestTypes.WORKSPACE_FOLDERS)
+    @send_response
```
Comment (ccordoba12) [inline spyder/plugins/editor/lsp/providers/workspace.py]: Please change the order of these decorators to make text easier to read, like this:

```python
@send_response
@handles(LSPRequestTypes.WORKSPACE_FOLDERS)
```

File: spyder/plugins/editor/lsp/providers/workspace.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        request_params = {
+            'added': added_folders,
+            'removed': removed_folders
+        }
+        if not workspace_settings['workspaceFolders']['supported']:
+            request_params[ClientConstants.CANCEL] = True
+        return request_params
+
+    @handles(LSPRequestTypes.WORKSPACE_CONFIGURATION)
+    @send_response
```
Comment (ccordoba12) [inline spyder/plugins/editor/lsp/providers/workspace.py]: Please also change the order of these decorators.

File: spyder/plugins/editor/lsp/transport/common/consumer.py [hunk 6]
```diff
@@ -124,6 +124,8 @@ def run(self):
     def parse_headers(self, headers):
         logger.debug(headers)
         headers = headers.split(b'\r\n')
+        headers = [header.split(b'\n')[-1] if b'\n' in header else header
+                   for header in headers]
```
Comment (ccordoba12) [inline spyder/plugins/editor/lsp/transport/common/consumer.py]: Why this change? And also, is the line above this one (i.e. `headers = headers.split(b'\r\n')`) necessary? Or can it be removed?

File: spyder/plugins/projects/plugin.py [hunk 7]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
             self.recent_projects.insert(0, project)
             self.recent_projects = self.recent_projects[:10]
 
+    def register_lsp_server_settings(self, settings):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Please add a docstring to this method.

File: spyder/plugins/projects/plugin.py [hunk 8]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
             self.recent_projects.insert(0, project)
             self.recent_projects = self.recent_projects[:10]
 
+    def register_lsp_server_settings(self, settings):
+        self.lsp_ready = True
+        if self.current_active_project:
+            path = self.get_active_project_path()
+            self.notify_project_open(path)
+
+    def stop_lsp_services(self):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 9]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+    def register_lsp_server_settings(self, settings):
+        self.lsp_ready = True
+        if self.current_active_project:
+            path = self.get_active_project_path()
+            self.notify_project_open(path)
+
+    def stop_lsp_services(self):
+        self.lsp_ready = False
+
+    def emit_request(self, method, params, requires_response):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 10]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+        params['requires_response'] = requires_response
+        params['response_instance'] = self
+        self.main.lspmanager.broadcast_request(method, params)
+
+    @Slot(str, dict)
+    def handle_response(self, method, params):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 11]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+        if method in self.handler_registry:
+            handler_name = self.handler_registry[method]
+            handler = getattr(self, handler_name)
+            handler(params)
+
+    @request(method=LSPRequestTypes.WORKSPACE_WATCHED_FILES_UPDATE,
+             requires_response=False)
     @Slot(str, str, bool)
     def file_moved(self, src_file, dest_file, is_dir):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 12]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+        entries = [addition_entry, deletion_entry]
+        params = {
+            'params': entries
+        }
+        return params
+
+    @request(method=LSPRequestTypes.WORKSPACE_WATCHED_FILES_UPDATE,
+             requires_response=False)
     @Slot(str, bool)
     def file_created(self, src_file, is_dir):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 13]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+                'file': src_file,
+                'kind': FileChangeType.CREATED
+            }]
+        }
+        return params
+
+    @request(method=LSPRequestTypes.WORKSPACE_WATCHED_FILES_UPDATE,
+             requires_response=False)
     @Slot(str, bool)
     def file_deleted(self, src_file, is_dir):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 14]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+                'file': src_file,
+                'kind': FileChangeType.DELETED
+            }]
+        }
+        return params
+
+    @request(method=LSPRequestTypes.WORKSPACE_WATCHED_FILES_UPDATE,
+             requires_response=False)
     @Slot(str, bool)
     def file_modified(self, src_file, is_dir):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 15]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+            'params': [{
+                'file': src_file,
+                'kind': FileChangeType.CHANGED
+            }]
+        }
+        return params
+
+    @request(method=LSPRequestTypes.WORKSPACE_FOLDERS_CHANGE,
+             requires_response=False)
+    def notify_project_open(self, path):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 16]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+        params = {
+            'folder': path,
+            'instance': self,
+            'kind': 'addition'
+        }
+        return params
+
+    @request(method=LSPRequestTypes.WORKSPACE_FOLDERS_CHANGE,
+             requires_response=False)
+    def notify_project_close(self, path):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

File: spyder/plugins/projects/plugin.py [hunk 17]
```diff
@@ -511,22 +517,124 @@ def add_to_recent(self, project):
...
+            'folder': path,
+            'instance': self,
+            'kind': 'deletion'
+        }
+        return params
+
+    @request(method=LSPRequestTypes.WORKSPACE_APPLY_EDIT,
+             requires_response=False)
+    @handles(LSPRequestTypes.WORKSPACE_APPLY_EDIT)
+    def handle_workspace_edit(self, params):
```
Comment (ccordoba12) [inline spyder/plugins/projects/plugin.py]: Add docstring.

**Comment by andfoy** [comment]
> You can start by testing: send_workspace_folders, send_workspace_folders_change, send_workspace_configuration and send_watched_files_change.

And please also add tests for notify_project_open, notify_project_close, and file_deleted for projects

I think most of the functionality cannot be tested because the pyls does not handle it and also because most of these calls are either requests posted by the server to be answered to the client, or notifications to the server, which do not have a response by the server.

**Comment by ccordoba12** [comment]
You're right. I thought we could be able to add tests for that. Then please address my comments and this one should be ready.
